java/ql/src/Security/CWE/CWE-927/ImplicitPendingIntents.qhelp

```diff
@@ -5,42 +5,42 @@
-  receiving application if they were not previously set. This means that a mutable <code>PendingIntent</code> that has 
-  not defined a destination component (that is, an implicit <code>PendingIntent</code>) can be altered to execute an 
+<p>A <code>PendingIntent</code> is used to wrap an <code>Intent</code> that will be supplied and executed by another
+  application. When the <code>Intent</code> is executed, it behaves as if it were run directly by the supplying
+  application, using the privileges of that application.</p>
+<p>If a <code>PendingIntent</code> is configured to be mutable, the fields of its internal <code>Intent</code> can be changed by the
+  receiving application if they were not previously set. This means that a mutable <code>PendingIntent</code> that has
+  not defined a destination component (that is, an implicit <code>PendingIntent</code>) can be altered to execute an
   arbitrary action with the privileges of the application that created it.</p>
-<p>If an implicit <code>PendingIntent</code> is obtainable by a malicious application by any of the following means:</p>
+<p>A malicious application can access an implicit <code>PendingIntent</code> as follows:</p>
 <ul>
-  <li>It is wrapped and sent as an extra of another implicit Intent</li>
-  <li>It is sent as the action of a Slide</li>
-  <li>It is sent as the action of a Notification</li>
+  <li>It is wrapped and sent as an extra of another implicit <code>Intent</code>.</li>
+  <li>It is sent as the action of a <code>Slide</code>.</li>
+  <li>It is sent as the action of a <code>Notification</code>.</li>
 </ul>
 <p></p>
-<p>the attacker could modify the underlying Intent and execute an arbitrary action with elevated privileges. 
-  This could give the malicious application access to private components of the victim application, 
-  or the ability to perform actions without having the necessary permissions.</p>
+<p>On gaining access, the attacker can modify the underlying <code>Intent</code> and execute an arbitrary action
+  with elevated privileges. This could give the malicious application access to private components of the victim
+  application, or the ability to perform actions without having the necessary permissions.</p>
 </overview>
 
 <recommendation>
-<p>Avoid creating implicit <code>PendingIntent</code>s. This means that the underlying Intent should always have an 
+<p>Avoid creating implicit <code>PendingIntent</code>s. This means that the underlying <code>Intent</code> should always have an 
   explicit destination component.</p>
-<p>Also, when adding the <code>PendingIntent</code> as an extra of another Intent, make sure that said Intent also has
+<p>Also, when adding the <code>PendingIntent</code> as an extra of another <code>Intent</code>, make sure that said <code>Intent</code> also has
 an explicit destination component, so that it is not delivered to untrusted applications.</p>
```

Comment:
Thanks for the explanation in: https://github.com/github/codeql/pull/6779#issuecomment-1018375595.

One thing that remains unclear is which `Intent` is meant by "said <code>Intent</code>". I think what we mean is that all `Intent`s should have an explicit destination component. So maybe:

```suggestion
<p>When you add the <code>PendingIntent</code> as an extra of another <code>Intent</code>, make sure that this second <code>Intent</code> also has
an explicit destination component, so that it is not delivered to untrusted applications.</p>
```

Reply:
Exactly. Good suggestion, thanks!